src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -37,6 +37,65 @@ void three_index_constraint(
     *constraint_it = *d_spacetime_metric_it - *phi_it;
   }
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> gauge_constraint(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  auto constraint =
+      make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(pi, 0.0);
+  gauge_constraint<SpatialDim, Frame, DataType>(
+      &constraint, gauge_function, spacetime_normal_one_form,
+      spacetime_normal_vector, inverse_spatial_metric,
+      inverse_spacetime_metric, pi, phi);
+  return constraint;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void gauge_constraint(
+    gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  *constraint = make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(pi, 0.0);
```

Comment:
Why do you need to do this if the allocation is already set? Doing this allocation completely defeats the purpose of the return-by-reference approach. You should check if the allocation is necessary and only do it if it is.